File: Java/DesignPatterns/Observer_design_pattern.java
// package DesignPatterns;

import java.util.ArrayList;
import java.util.List;

interface Subscriber{
    void notifyMe(String videoId);
}

class Ananya implements Subscriber{
    @Override
    public void notifyMe(String videoId) {
        System.out.println("Ananaya: New Video uploaded"+videoId);
    }
}

class Amol implements Subscriber{
    @Override
    public void notifyMe(String videoId) {
        System.out.println("Amol: New Video uploaded"+videoId);

    }
}


class YoutubeChannel{
    String name;
    List<Subscriber> subscribers;
    public YoutubeChannel(){
        subscribers = new ArrayList<>();
    }
    void addSubsriber(Subscriber sub){
        subscribers.add(sub);
    }

    void uploadVideo(String videoId){
        //10000 linee of code
    notifySubscriber(videoId);
    }
    void notifySubscriber(String videoId){
        for(Subscriber subscriber : subscribers){
            subscriber.notifyMe(videoId);
        }
    }

}


public class Observer_design_pattern {
    public static void main(String[] args) {
        YoutubeChannel channel = new YoutubeChannel();
        channel.addSubsriber(new Amol());
        channel.addSubsriber(new Ananya());

        channel.uploadVideo("how to avoid layoff in your company");
    }
}

File: Golang/DesignPatterns/Observer_design_pattern.go
package main

import "fmt"

type Subscriber interface {
	NotifyMe(videoId string)
}

type Ananya struct{}

func (a *Ananya) NotifyMe(videoId string) {
	fmt.Printf("Ananaya: New Video uploaded%s\n", videoId)
}

type Amol struct{}

func (a *Amol) NotifyMe(videoId string) {
	fmt.Printf("Amol: New Video uploaded%s\n", videoId)
}

type YoutubeChannel struct {
	name        string
	subscribers []Subscriber
}

func (c *YoutubeChannel) addSubsriber(sub Subscriber) {
	c.subscribers = append(c.subscribers, sub)
}

func (c *YoutubeChannel) uploadVideo(videoId string) {
	// 10000 linee of code
	c.notifySubscriber(videoId)
}

func (c *YoutubeChannel) notifySubscriber(videoId string) {
	for _, subscriber := range c.subscribers {
		subscriber.NotifyMe(videoId)
	}
}

func main() {
	channel := &YoutubeChannel{}
	channel.addSubsriber(&Amol{})
	channel.addSubsriber(&Ananya{})

	channel.uploadVideo("how to avoid layoff in your company")
}
